### PR TITLE
fix: style relation checker using substance variables

### PR DESCRIPTION
**Pull request by liangyiliang**
# Description

Related issue/PR: #1123

This PR fixes the uncaught error in the Style compiler when selector relations refer to a variable not declared in the selector, but exists in the substance program. It now throws an error. We added a test case on this error. Finally, we improved upon how the "suggested" variables in the error message distinguish between Style and Substance variables.

# Implementation strategy and design decisions

The style compiler checks through the relations in the `where` clause, primarily using functions within the Substance checker, which checks the consistency of expressions with respect to a `SubstanceEnv`. In particular, it checks that each variable referred in the expressions are declared inside `SubstanceEnv`.

In order to call the Substance checker, we would need to provide a `SubstanceEnv` that contains all the variables that we want the relations to be able to access. Previously, we just take the current `SubstanceEnv` (as used in the Substance program), and add all the variables declared in the style header. The problem is that now this `SubstanceEnv` contains both variables declared in the style header (which we want to be able to access) and variables not declared in the style header but exists in the substance program.

The solution is simply to create a new, empty `SubstanceEnv` and add in the variables declared in the style header.

# Examples with steps to reproduce them

[This](https://merging-envs.penrose-72l.pages.dev/try/?gist=a7a33751a606a5688cb7c7cdf7aea9e7) is the same example as used in the original issue, running in the version in this PR. Previously, this did not trigger an error, and only fails silently under the browser console. Now it does trigger a sensible error.

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new ESLint warnings
- [x] I have reviewed any generated registry diagram changes

### CODE REVIEW

**Comment by samestep** [comment]
Just to clarify, does #1123 become fixed via this PR? If so, could you modify your PR description autoclose it using GitHub's special syntax? https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

**Comment by liangyiliang** [comment]
> Just to clarify, does #1123 become fixed via this PR? If so, could you modify your PR description autoclose it using GitHub's special syntax? https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

Done!

**Comment by samestep** [comment]
> Done!

Thanks! And in future you can do this by just saying "Closes #1123" instead of "Related issue/PR: #1123". Actually, @wodeni do you think we should change this in `.github/PULL_REQUEST_TEMPLATE.md`? I've seen a bunch of people get tripped up by this in the past.

**Review by wodeni — APPROVED** [review]

File: packages/core/src/compiler/Style.ts [hunk 1]
```diff
@@ -567,22 +567,31 @@ const mergeMapping = (
...
+          ...varEnv.varIDs,
+        ],
+      };
+    case "SubVar":
+      return {
+        ...varEnv,
+        vars: varEnv.vars.set(
+          bindingForm.contents.value,
+          toSubstanceType(styType)
+        ),
```
Comment (samestep) [inline packages/core/src/compiler/Style.ts:589]: This code looks duplicated from the above code for `StyVar`; does it make sense to pull it to above the `switch`?

**Review by samestep — APPROVED** [review]
Looks great, thanks Leo! Just one minor comment.

File: packages/core/src/compiler/Style.ts (hunk 1 above)
Comment (liangyiliang) [reply packages/core/src/compiler/Style.ts]: Done!

File: packages/core/src/compiler/Style.ts [hunk 2]
```diff
@@ -567,22 +567,29 @@ const mergeMapping = (
...
-    case "StyVar": {
-      // G || (y : |T) |-> G[y : T] (shadowing any existing Sub vars)
+    case "StyVar":
       return {
         ...varEnv,
-        vars: varEnv.vars.set(
-          bindingForm.contents.value,
-          toSubstanceType(styType)
-        ),
+        vars: vars,
```
Comment (samestep) [inline packages/core/src/compiler/Style.ts:578]:
```suggestion
        vars,
```

File: packages/core/src/compiler/Style.ts [hunk 3]
```diff
@@ -567,22 +567,29 @@ const mergeMapping = (
...
+        varIDs: [
+          dummyIdentifier(bindingForm.contents.value, "Style"),
+          ...varEnv.varIDs,
+        ],
+      };
+    case "SubVar":
+      return {
+        ...varEnv,
+        vars: vars,
```
Comment (samestep) [inline packages/core/src/compiler/Style.ts:587]:
```suggestion
        vars,
```

**Pull request merged**